### ISSUE TITLE
Switching heart rate monitor to visible

### ISSUE THREAD

**Issue by desperadoduck** [Freeyourgadget/Gadgetbridge#1606]
#### Your issue is:

In the MiFit-App the device can be switched to a visible mode, which allows other apps (like e.g. Locus Maps) to connect to the device and record the heart rate.
This setting seems to be missing from the gadgetbridge app (you can switch it by using MiFit, and then use it afterwards on a device which only has GadgetBridge installed, so it seems to be a setting which is stored on the device).

#### Your wearable device is:

Mi Smart band 4
FW: 1.0.5.66

#### Your android version is:
8

#### Your Gadgetbridge version is:
0.35.2

**Comment by desperadoduck** [comment]
oops, there are two devices involved, so let me rephrase the last part to make it more clear:
you can switch it by using MiFit, and then use it afterwards on a phone which only has GadgetBridge installed, so it seems to be a setting which is stored on the Mi Smart Band.

**Comment by LuccoJ** [comment]
Isn't this also a major privacy issue, if anything (not just that particular phone, possibly...?) can read information from the watch?

**Comment by desperadoduck** [comment]
Yes, I think it probably might be. On the other hand the sensor is only visible when the mi band is in training mode. Therefore the privacy risks depends on your usecase: if you use this feature when hiking in the mountains, someone would have to follow you close by for a long while to get a significant amount of heart beat data, which I don't think is very probable. In a gym, it would be a completely different case... I'm also not sure if more than one device can connect at the same time to the sensor - if thats not the case, it would reduce the risk further, if you actively use this feature. 
But that's why I'd like to have a switch for that: I'd like to turn it on exactly when I need it and when I think the risk is managable, and turn it off, when I don't. There probably should be some kind of warning message when turning that on, though.

**Comment by vanous** [comment]
I don't think this is breaching user privacy, as long as the user selects this. Actually, it would be cool to do, there are other floss apps which could utilize that, see for example here this issue/commits: https://github.com/bailuk/AAT/issues/41

**Comment by ashimokawa** [comment]
@desperadoduck 
Does it work when Gadgetbridge is not connected to the Band?

**Comment by desperadoduck** [comment]
OK, I tried it:
1. If Gadgetbridge is disconnected, the other app reports "invalid data" for the heart rate sensor. As soon as I connect Gadgetbridge, valid data arrives.
2. If Gadgetbridge is connected, but the other app runs on a second phone, the app sees the mi band, but seemingly cannot connect to it.

So to summarize everything we know at the moment: to use this feature it looks like you need to
a) connect gadgetbridge
b) run the other other app on the same device where gadgetbridge is running
c) run a training on the band and
d) have the sensor configured to be visible (which is only possible via MiFit at the moment)
If one of these points does not apply, it doesn't seem to be working.

**Comment by ashimokawa** [comment]
@desperadoduck 
Would it help you if I do d) ?

**Comment by desperadoduck** [comment]
@ashimokawa 
Of course it would :-) Then I wouldn't need to keep a second phone around just to run MiFit... ;-)

**Comment by vanous** [comment]
@desperadoduck could you please describe/post screenshot where in MiFit you can enable this option?

Thank you

**Comment by desperadoduck** [comment]
Of course. My phone language is German, so whenever there is text I write the original German text and in brackets what I think might be the English translation:

1. Start MiFit
2. Click on "Profil" (profile)
3. Click on "Meine Geräte" -> "Mi INTELLI-Band-4" (My Devices -> Mi INTELLI-Band-4)
4. Scroll down a bit. There are some options with icons, and after that there are options without Icons. The first one without icon should be "band finden" ("search band"). The next option on this screen is "Erkennbar" (visible/recognizable). This is the option we are looking for. When you click on this option you get to a screen which says "Sichtbarkeit. Aktivierung dieser Funktion kann ein wenig die Akkulaufzeit verringern" ("Visibility. Activation of this function may reduce your battery life a bit").

I hope you find it that way, otherwise I can look if I can switch the language somehow.

**Comment by vanous** [comment]
Thank you. I do not have MiFit, so i can only try probably next week with some friend, but i have though that this only makes it visible over bluetooth as in normal bluetooth visibility... interesting.

**Comment by vanous** [comment]
Also, it seems MiFit4 is needed for this...

**Comment by vanous** [comment]
Cool. I can translate the screenshots if needed

**Comment by ashimokawa** [comment]
@desperadoduck 
Got this working as a dirty test (not yet commited) on the Mi Band 3 (assume the same works for Mi Band 4)

Just didn't know about it before since I only use Gadgetbridge and never Mi Fit.
I didn't need any special app to test, the generic BLExplorer from F-Droid did the job for testing if HR data arrives.

**Comment by ashimokawa** [comment]
Done, confirmed working on Bip, Mi Band 3
Assumed working on Cor and Mi Band 4.
Will check on Mi Band 2.

HR data is also sent when in "Status" on the Bip, no need to start an activity

**Comment by vanous** [comment]
this is very cool!

**Comment by vanous** [comment]
How do I confirm this, should i able to see the HR data in BLE explorer? When I select heart rate there,  value is always 0=0...

**Comment by vanous** [comment]
characteristic UUID: 00002a39-0000-1000-8000-00805f9b34fb service UUID: 0000180d-0000-1000-8000-00805f9b34fb value: 0 = 0 =

**Comment by desperadoduck** [comment]
@ashimokawa: I can confirm that it's working with my Mi Band 4. This is really great!

@vanous: BLE explorer crashes for me, so I can't comment directly on that. Did you try the steps that I described earlier? I updated them a bit:
a) connect gadgetbridge
b) configure the sensor to be visible
c) start a training on the mi band
d) run the other other app on the same device where gadgetbridge is running
e) the other app should show live heart rate data (it might take some seconds for the first measurment to be valid - you can check on the display of your band if it has valid heart rate data).

**Comment by vanous** [comment]
>BLE explorer crashes for me

:(

Yes, i could see the data in the LocusMap (i made a purchase, tested this and then returned the app, as i didn't like it :) :))

For me, i will not use this very much, because the Bip has GPS, but i would like to see AAT working with our data. Right now it is not and i am not sure why... can you perhaps try AAT?  At first it looks strange, but it is designed for sunny day on a bike, and the design makes sense there. 

In settings, search for BLE devices, then connect, then run tracking and open up dashboard to see the data...
https://f-droid.org/packages/ch.bailu.aat/

**Comment by desperadoduck** [comment]
Interesting... I can't see the heart rate in AAT. I would have expected that the sensor works in any app, after activating it. Maybe there are different versions of the standard, and not all apps support both?

**Comment by vanous** [comment]
here is their issue about MB3: https://github.com/bailuk/AAT/issues/62 , perhaps we can hep to troubleshoot this.

**Comment by vanous** [comment]
For the record, once the sharing is enabled and activity is started on the device (or status screen on Bip is used) live data is visible in BLE Explorer, in the Heartrate section, first line (the 00002a37... characteristics). You must enable the notify toggle. GB does not need to run nor be connected anymore, the sensor remains available.

**Comment by ashimokawa** [comment]
About the AAT problem - they should should parse the "incompatible data format" - it is just the raw heart rate, no problem to fix that on their side IMO 🍠

**Comment by vanous** [comment]
Agree. Btw, i added this to [wiki](https://codeberg.org/Freeyourgadget/Gadgetbridge/wiki/Heart-Rate-data-sharing), linked right now to MB3,MB4, Bip. 
If anyone finds/confirms compatibility with other devices, please edit or report.

**Comment by ashimokawa** [comment]
@vanous 
Thanks. I think it is pretty much safe to assume it works with Amazfit Cor/Cor 2 also.
Will try Mi Band 2 today.
